README
SysBench:  System evaluation benchmark

The idea is to quickly get an impression about system performance for MySQL
usage without setting up complex benchmark and even without installing MySQL.
In some cases this is very helpful. This is also the reason for having
everything in simple file not depending on any external libraries.

Description:
This benchmark was designed for identifying basic system parameters, as they
are important for system using MySQL (w Innodb) under intensive load.
Handling of IO in case of many parallel requests, checked as well as 
memory allocation/transfer speed and scheduler performance. 
CPU is benchmarked by using 64bit integer manipulation using Euklid 
algorithms for prime number computation. 
Benchmarks are designed to show benefit of multiple CPUs as well as of
multiple hard drives, battery backed up write cache.

Sequential read/writes test is a bit tricky, it means request are just given
out sequentially, but there is no guaranty they will be actually executed this
way, it could result in some sort of random IO in case of bad IO or threads
scheduler

Installing:
See the file INSTALL for generic installation instructions.
In most cases you will need the following steps:
    ./autogen.sh (unnecessary if you are building from a release tarball)
    ./configure (or ./configure --without-mysql to compile w/o MySQL support)
    make
    make install

Running:
The general syntax is
    sysbench [<general-options>] [<test-options>] <command>
To see available options, tests and commands type:
    sysbench help
or
    sysbench --test=<name> help
to see options supported by a specified test. Default values for the options
are specified in brackets after description.
Currently supported commands are help, run and create-db (for the MySQL test).
The latter will create a table and fill it with values required to run the 
MySQL test, using database, table name and size passed as MySQL test options. 

Notes:
* Make sure number of threads is at least 2 times of numbers CPUs you have.
* For the fileio test make sure you're using file size which is at least 5 
times larger than your memory for reliable results. Make sure to use enough 
files to keep size of each less than 2/4G. 
* For the threads test --thread-locks should be at least 2 times less than 
number of threads.
* If you're running on 64bit CPU make sure to compile 64bit binary for good
results. 


Basic design:
This test is designed to be easily extensible to add your own test operations,
as well as create operation mixes to benchmark some particular load.
The architecture is very simple. Several threads are started and they all
run in the loop asking for requests from get_request() function,  until
NULL operation is received. For each request execute_request() is executed,
which runs it as designed. This allows to design test highly scalable as well
as producing large amount of conflicts if needed for test case.

Bug reporting:
Please report any bugs you encounter to <sysbench-developers@lists.launchpad.net>. 


ChangeLog
2020-04-24  Alexey Kopytov  <akopytov@gmail.com>

	* version 1.0.20
	* build/CI/packaging: Add arm64 to Travis CI matrix (#358)
	* build/CI/packaging: add Ubuntu Focal
	* build/CI/packaging: remove Fedora Rawhide from CI matrix
	* build/CI/packaging: fix regression tests to work with MySQL 8.0.19+
	* build/CI/packaging: fix macOS builds in Travis
	* build/CI/packaging: remove Ubuntu Disco (EOL)

2019-12-08  Alexey Kopytov  <akopytov@gmail.com>

	* version 1.0.19
	* build/CI/packaging: fix Ubuntu packaging for Bionic and later versions
	* regression tests: compatibility fix for PostgreSQL 12
	* build/CI/packaging: fix macOs builds in Travis
	* build/CI/packaging: add Fedora 31.

2019-10-21  Alexey Kopytov  <akopytov@gmail.com>

	* version 1.0.18
	* build/CI/packaging: add Ubuntu Eoan.
	* build/CI/packaging: remove Ubuntu Cosmic (EOL).
	* build/CI/packaging: add CentOS 8.
	* build/CI/packaging: add Ubuntu Disco.
	* build/CI/packaging: remove Ubuntu Trusty (EOL).
	* build/CI/packaging: remove Fedora 28 (EOL).
	* build/CI/packaging: add Fedora 30.
	* build/CI/packaging: cherry-pick fix for LuaJIT/LuaJIT#484 to
	fix builds on macOS Mojave.
	* build/CI/packaging: add Debian Buster

2019-03-15  Alexey Kopytov  <akopytov@gmail.com>

	* version 1.0.17
	* build/CI/packaging: update RPM spec to support RHEL8-beta
	(thanks to Alexey Bychko for the patch)
	* regression tests: remove unnecessary error leading to opt_rate.t instability.
	* --rate mode: return a non-zero exit code on event queue
	overflow.
	* --rate mode: fix a bogus error about eventgen thread termination

2018-12-16  Alexey Kopytov  <akopytov@gmail.com>

	* version 1.0.16
	* build/CI/packaging: add Ubuntu Cosmic.
	* build/CI/packaging: add Fedora 29.
	* build/CI/packaging: remove Fedora 27 (EOL).
	* SQL API: fix GH-282 (Mysql's fetch_row() is broken)
	* --rate mode: fix latency stats skew on low rates
	* Lua: Add /usr/share/lua/5.1 to LUA_PATH and /usr/lib/lua/5.1
	to LUA_CPATH.
	* build/CI/packaging: add -Wvla to default compiler flags.
	* build/CI/packaging: fix debian/changelog format
	* build/CI/packaging: fix buildpack.sh to not push multiple file
	types to packagecloud.
	* build/CI/packaging: add libaio-dev to Debian/Ubuntu build
	dependencies.

2018-07-03  Alexey Kopytov  <akopytov@gmail.com>

	* version 1.0.15
	* CI/build/packaging: add Fedora 28
	* CI/build/packaging: add Ubuntu Bionic
	* CI/build/packaging: remove Fedora 26 (EOL)
	* CI/build/packaging: remove Debian Wheezy (EOL)
	* fileio: fix GH-229 (--file-fsync-freq=0 seems to prevent
	fsync() at the end of the test)
	* command line: improve parsing of boolean command line options
	* tests: fix GH-220 (Testsuite api_sql_mysql.t failed ...)
	* tests: fix GH-223 (test failure on ppc64)
	* tests: fix opt_help.t to pass when the binary is not
	configured with MySQL support
	* MySQL driver: use it by default in DB benchmarks

2018-04-01  Alexey Kopytov  <akopytov@gmail.com>

	* version 1.0.14
	* reports: fix JSON stats reporter to produce valid JSON
	(GH-195)
	* Lua SQL API: don't crash when query_row() is called with a
	SELECT returning empty result set
	* Lua SQL API: don't crash when bulk insert API calls are used
	out of order
	* regression tests: make PostgreSQL tests compatible with the
	new dump format introduced in 10.3
	* regression tests: minor stability and coverage improvements

2018-02-17  Alexey Kopytov  <akopytov@gmail.com>

	* version 1.0.13
	* remove Ubuntu Zesty from CI/build/packaging matrices (EOL)
	* minor cleanups in build scripts
	* improve report formatting for long latency values
	* fileio: --file-extra-flags now accepts a list of flags rather
	than just a single value
	* OLTP: re-prepare prepared statements after reconnects, i.e. in
	cases when a server connection is lost and sysbench is
	configured to ignore such errors

2018-01-17  Alexey Kopytov  <akopytov@gmail.com>

	* version 1.0.12
	* improve --rate mode precision for high argument values
	* add Fedora Rawhide and Debian Sid to CI matrix
	* fix compile-time architecture detection for some Broadwell
	CPUs which were incorrectly identified as Core 2.
	* remove build dependency on xxd (and vim-minimal package)
	* fix Lua API to correctly stop the benchmark when event()
	returns a value other than nil or false (thanks to caojiafeng
	for the patch)
	* fix the fileio benchmark when the specified file size is not a
	multiple of block size
	* fix the fileio benchmark to throw a descriptive error when the
	specified file size does not match the size of files created by
	'prepare'
	* remove Fedora 25 from CI/build/packaging matrices (EOL)
	* minor improvements in tests and documentation.

2017-12-09  Alexey Kopytov  <akopytov@gmail.com>

	* version 1.0.11
	* add Debian Stretch to CI/build/packaging matrices
	* add Fedora 27 to CI/build/packaging matrices
	* make statistic counters usable from Lua scripts
	* fix the PostgreSQL driver to be compatible with CockroachDB
	(GH-180)
	* fix oltp_insert.lua to work correctly when both --tables and
	--threads are greater than 1 (GH-178)
	* fix FreeBSD builds by adding -rdynamic to the default linker
	flags (GH-174)
	* minor documentation updates

2017-10-25  Alexey Kopytov  <akopytov@gmail.com>

	* version 1.0.10
	* fixed PK conflicts in oltp_insert.lua by creating empty tables
	on 'prepare'
	* made sysbench.opt available to init()/done() by exporting it
	to the global Lua state
	* added Fedora 26 (both x86_64 and AArch64) to the list of
	supported and tested distributions
	* fixed GH-172: sysbench 1.0.9 doesn't build with mariadb 10.2.8
	* add the /usr/local LuaRocks root directory to default LUA_PATH
	and LUA_CPATH
	* removed Fedora 24, Ubuntu Precise, Yakkety from default build
	matrices
	* added Ubuntu Artful to default build matrices

2017-09-05  Alexey Kopytov  <akopytov@gmail.com>

	* version 1.0.9
	* fixed oltp_delete.lua to not use INSERT statements for
	consistency with other oltp_* benchmarks (GH-168)

src/lj_arch.h
/*
** Target architecture selection.
** Copyright (C) 2005-2015 Mike Pall. See Copyright Notice in luajit.h
*/

#ifndef _LJ_ARCH_H
#define _LJ_ARCH_H

#include "lua.h"

/* Target endianess. */
#define LUAJIT_LE	0
#define LUAJIT_BE	1

/* Target architectures. */
#define LUAJIT_ARCH_X86		1
#define LUAJIT_ARCH_x86		1
#define LUAJIT_ARCH_X64		2
#define LUAJIT_ARCH_x64		2
#define LUAJIT_ARCH_ARM		3
#define LUAJIT_ARCH_arm		3
#define LUAJIT_ARCH_ARM64	4
#define LUAJIT_ARCH_arm64	4
#define LUAJIT_ARCH_PPC		5
#define LUAJIT_ARCH_ppc		5
#define LUAJIT_ARCH_MIPS	6
#define LUAJIT_ARCH_mips	6

/* Target OS. */
#define LUAJIT_OS_OTHER		0
#define LUAJIT_OS_WINDOWS	1
#define LUAJIT_OS_LINUX		2
#define LUAJIT_OS_OSX		3
#define LUAJIT_OS_BSD		4
#define LUAJIT_OS_POSIX		5

/* Select native target if no target defined. */
#ifndef LUAJIT_TARGET

#if defined(__i386) || defined(__i386__) || defined(_M_IX86)
#define LUAJIT_TARGET	LUAJIT_ARCH_X86
#elif defined(__x86_64__) || defined(__x86_64) || defined(_M_X64) || defined(_M_AMD64)
#define LUAJIT_TARGET	LUAJIT_ARCH_X64
#elif defined(__arm__) || defined(__arm) || defined(__ARM__) || defined(__ARM)
#define LUAJIT_TARGET	LUAJIT_ARCH_ARM
#elif defined(__aarch64__)
#define LUAJIT_TARGET	LUAJIT_ARCH_ARM64
#elif defined(__ppc__) || defined(__ppc) || defined(__PPC__) || defined(__PPC) || defined(__powerpc__) || defined(__powerpc) || defined(__POWERPC__) || defined(__POWERPC) || defined(_M_PPC)
#define LUAJIT_TARGET	LUAJIT_ARCH_PPC
#elif defined(__mips__) || defined(__mips) || defined(__MIPS__) || defined(__MIPS)
#define LUAJIT_TARGET	LUAJIT_ARCH_MIPS
#else
#error "No support for this architecture (yet)"
#endif

#endif

/* Select native OS if no target OS defined. */
#ifndef LUAJIT_OS

#if defined(_WIN32) && !defined(_XBOX_VER)
#define LUAJIT_OS	LUAJIT_OS_WINDOWS
#elif defined(__linux__)
#define LUAJIT_OS	LUAJIT_OS_LINUX
#elif defined(__MACH__) && defined(__APPLE__)
#define LUAJIT_OS	LUAJIT_OS_OSX
#elif (defined(__FreeBSD__) || defined(__FreeBSD_kernel__) || \
       defined(__NetBSD__) || defined(__OpenBSD__) || \
       defined(__DragonFly__)) && !defined(__ORBIS__)
#define LUAJIT_OS	LUAJIT_OS_BSD
#elif (defined(__sun__) && defined(__svr4__)) || defined(__CYGWIN__)
#define LUAJIT_OS	LUAJIT_OS_POSIX
#else
#define LUAJIT_OS	LUAJIT_OS_OTHER
#endif

#endif

/* Set target OS properties. */
#if LUAJIT_OS == LUAJIT_OS_WINDOWS
#define LJ_OS_NAME	"Windows"
#elif LUAJIT_OS == LUAJIT_OS_LINUX
#define LJ_OS_NAME	"Linux"
#elif LUAJIT_OS == LUAJIT_OS_OSX
#define LJ_OS_NAME	"OSX"
#elif LUAJIT_OS == LUAJIT_OS_BSD
#define LJ_OS_NAME	"BSD"
#elif LUAJIT_OS == LUAJIT_OS_POSIX
#define LJ_OS_NAME	"POSIX"
#else
#define LJ_OS_NAME	"Other"
#endif

#define LJ_TARGET_WINDOWS	(LUAJIT_OS == LUAJIT_OS_WINDOWS)
#define LJ_TARGET_LINUX		(LUAJIT_OS == LUAJIT_OS_LINUX)
#define LJ_TARGET_OSX		(LUAJIT_OS == LUAJIT_OS_OSX)
#define LJ_TARGET_IOS		(LJ_TARGET_OSX && (LUAJIT_TARGET == LUAJIT_ARCH_ARM || LUAJIT_TARGET == LUAJIT_ARCH_ARM64))
#define LJ_TARGET_POSIX		(LUAJIT_OS > LUAJIT_OS_WINDOWS)
#define LJ_TARGET_DLOPEN	LJ_TARGET_POSIX

#ifdef __CELLOS_LV2__
#define LJ_TARGET_PS3		1
#define LJ_TARGET_CONSOLE	1
#endif

#ifdef __ORBIS__
#define LJ_TARGET_PS4		1
#define LJ_TARGET_CONSOLE	1
#undef NULL
#define NULL ((void*)0)
#endif

#ifdef __psp2__
#define LJ_TARGET_PSVITA	1
#define LJ_TARGET_CONSOLE	1
#endif

#if _XBOX_VER >= 200
#define LJ_TARGET_XBOX360	1
#define LJ_TARGET_CONSOLE	1
#endif

#ifdef _DURANGO
#define LJ_TARGET_XBOXONE	1
#define LJ_TARGET_CONSOLE	1
#define LJ_TARGET_GC64		1
#endif

#define LJ_NUMMODE_SINGLE	0	/* Single-number mode only. */
#define LJ_NUMMODE_SINGLE_DUAL	1	/* Default to single-number mode. */
#define LJ_NUMMODE_DUAL		2	/* Dual-number mode only. */
#define LJ_NUMMODE_DUAL_SINGLE	3	/* Default to dual-number mode. */

/* Set target architecture properties. */
#if LUAJIT_TARGET == LUAJIT_ARCH_X86

#define LJ_ARCH_NAME		"x86"
#define LJ_ARCH_BITS		32
#define LJ_ARCH_ENDIAN		LUAJIT_LE
#if LJ_TARGET_WINDOWS || __CYGWIN__
#define LJ_ABI_WIN		1
#else
#define LJ_ABI_WIN		0
#endif
#define LJ_TARGET_X86		1
#define LJ_TARGET_X86ORX64	1
#define LJ_TARGET_EHRETREG	0
#define LJ_TARGET_MASKSHIFT	1
#define LJ_TARGET_MASKROT	1
#define LJ_TARGET_UNALIGNED	1
#define LJ_ARCH_NUMMODE		LJ_NUMMODE_SINGLE_DUAL

#elif LUAJIT_TARGET == LUAJIT_ARCH_X64

#define LJ_ARCH_NAME		"x64"
#define LJ_ARCH_BITS		64
#define LJ_ARCH_ENDIAN		LUAJIT_LE
#if LJ_TARGET_WINDOWS || __CYGWIN__
#define LJ_ABI_WIN		1
#else
#define LJ_ABI_WIN		0
#endif
#define LJ_TARGET_X64		1
#define LJ_TARGET_X86ORX64	1
#define LJ_TARGET_EHRETREG	0
#define LJ_TARGET_JUMPRANGE	31	/* +-2^31 = +-2GB */
#define LJ_TARGET_MASKSHIFT	1
#define LJ_TARGET_MASKROT	1
#define LJ_TARGET_UNALIGNED	1
#define LJ_ARCH_NUMMODE		LJ_NUMMODE_SINGLE_DUAL
#ifdef LUAJIT_ENABLE_GC64
#define LJ_TARGET_GC64		1
#endif

#elif LUAJIT_TARGET == LUAJIT_ARCH_ARM

#define LJ_ARCH_NAME		"arm"
#define LJ_ARCH_BITS		32
#define LJ_ARCH_ENDIAN		LUAJIT_LE
#if !defined(LJ_ARCH_HASFPU) && __SOFTFP__
#define LJ_ARCH_HASFPU		0
#endif
#if !defined(LJ_ABI_SOFTFP) && !__ARM_PCS_VFP
#define LJ_ABI_SOFTFP		1
#endif
#define LJ_ABI_EABI		1
#define LJ_TARGET_ARM		1
#define LJ_TARGET_EHRETREG	0
#define LJ_TARGET_JUMPRANGE	25	/* +-2^25 = +-32MB */
#define LJ_TARGET_MASKSHIFT	0
#define LJ_TARGET_MASKROT	1
#define LJ_TARGET_UNIFYROT	2	/* Want only IR_BROR. */
#define LJ_ARCH_NUMMODE		LJ_NUMMODE_DUAL

#if __ARM_ARCH____ARM_ARCH_8__ || __ARM_ARCH_8A__
#define LJ_ARCH_VERSION		80
#elif __ARM_ARCH_7__ || __ARM_ARCH_7A__ || __ARM_ARCH_7R__ || __ARM_ARCH_7S__ || __ARM_ARCH_7VE__
#define LJ_ARCH_VERSION		70
#elif __ARM_ARCH_6T2__
#define LJ_ARCH_VERSION		61
#elif __ARM_ARCH_6__ || __ARM_ARCH_6J__ || __ARM_ARCH_6K__ || __ARM_ARCH_6Z__ || __ARM_ARCH_6ZK__
#define LJ_ARCH_VERSION		60
#else
#define LJ_ARCH_VERSION		50
#endif

#elif LUAJIT_TARGET == LUAJIT_ARCH_ARM64

#define LJ_ARCH_NAME		"arm64"
#define LJ_ARCH_BITS		64
#define LJ_ARCH_ENDIAN		LUAJIT_LE
#define LJ_TARGET_ARM64		1
#define LJ_TARGET_EHRETREG	0
#define LJ_TARGET_JUMPRANGE	27	/* +-2^27 = +-128MB */
#define LJ_TARGET_MASKSHIFT	1
#define LJ_TARGET_MASKROT	1
#define LJ_TARGET_UNIFYROT	2	/* Want only IR_BROR. */
#define LJ_TARGET_GC64		1
#define LJ_ARCH_NUMMODE		LJ_NUMMODE_DUAL
#define LJ_ARCH_NOJIT		1	/* NYI */

#define LJ_ARCH_VERSION		80

#elif LUAJIT_TARGET == LUAJIT_ARCH_PPC

#ifndef LJ_ARCH_ENDIAN
#if __BYTE_ORDER__ != __ORDER_BIG_ENDIAN__
#define LJ_ARCH_ENDIAN		LUAJIT_LE
#else
#define LJ_ARCH_ENDIAN		LUAJIT_BE
#endif
#endif

#if _LP64
#define LJ_ARCH_BITS		64
#if LJ_ARCH_ENDIAN == LUAJIT_LE
#define LJ_ARCH_NAME		"ppc64le"
#else
#define LJ_ARCH_NAME		"ppc64"
#endif
#else
#define LJ_ARCH_BITS		32
#define LJ_ARCH_NAME		"ppc"
#endif

#define LJ_TARGET_PPC		1
#define LJ_TARGET_EHRETREG	3
#define LJ_TARGET_JUMPRANGE	25	/* +-2^25 = +-32MB */
#define LJ_TARGET_MASKSHIFT	0
#define LJ_TARGET_MASKROT	1
#define LJ_TARGET_UNIFYROT	1	/* Want only IR_BROL. */
#define LJ_ARCH_NUMMODE		LJ_NUMMODE_DUAL_SINGLE

#if LJ_TARGET_CONSOLE
#define LJ_ARCH_PPC32ON64	1
#define LJ_ARCH_NOFFI		1
#elif LJ_ARCH_BITS == 64
#define LJ_ARCH_PPC64		1
#define LJ_TARGET_GC64		1
#define LJ_ARCH_NOJIT		1	/* NYI */
#endif

#if _ARCH_PWR7
#define LJ_ARCH_VERSION		70
#elif _ARCH_PWR6
#define LJ_ARCH_VERSION		60
#elif _ARCH_PWR5X
#define LJ_ARCH_VERSION		51
#elif _ARCH_PWR5
#define LJ_ARCH_VERSION		50
#elif _ARCH_PWR4
#define LJ_ARCH_VERSION		40
#else
#define LJ_ARCH_VERSION		0
#endif
#if _ARCH_PPCSQ
#define LJ_ARCH_SQRT		1
#endif
#if _ARCH_PWR5X
#define LJ_ARCH_ROUND		1
#endif
#if __PPU__
#define LJ_ARCH_CELL		1
#endif
#if LJ_TARGET_XBOX360
#define LJ_ARCH_XENON		1
#endif

#elif LUAJIT_TARGET == LUAJIT_ARCH_MIPS

#if defined(__MIPSEL__) || defined(__MIPSEL) || defined(_MIPSEL)
#define LJ_ARCH_NAME		"mipsel"
#define LJ_ARCH_ENDIAN		LUAJIT_LE
#else
#define LJ_ARCH_NAME		"mips"
#define LJ_ARCH_ENDIAN		LUAJIT_BE
#endif

#if !defined(LJ_ARCH_HASFPU)
#ifdef __mips_soft_float
#define LJ_ARCH_HASFPU		0
#else
#define LJ_ARCH_HASFPU		1
#endif
#endif

/* Temporarily disable features until the code has been merged. */
#if !defined(LUAJIT_NO_UNWIND) && __GNU_COMPACT_EH__
#define LUAJIT_NO_UNWIND	1
#endif

#if !defined(LJ_ABI_SOFTFP)
#ifdef __mips_soft_float
#define LJ_ABI_SOFTFP		1
#else
#define LJ_ABI_SOFTFP		0
#endif
#endif

#define LJ_ARCH_BITS		32
#define LJ_TARGET_MIPS		1
#define LJ_TARGET_EHRETREG	4
#define LJ_TARGET_JUMPRANGE	27	/* 2*2^27 = 256MB-aligned region */
#define LJ_TARGET_MASKSHIFT	1
#define LJ_TARGET_MASKROT	1
#define LJ_TARGET_UNIFYROT	2	/* Want only IR_BROR. */
#define LJ_ARCH_NUMMODE		LJ_NUMMODE_DUAL

#if _MIPS_ARCH_MIPS32R2
#define LJ_ARCH_VERSION		20
#else
#define LJ_ARCH_VERSION		10
#endif

#else
#error "No target architecture defined"
#endif

#ifndef LJ_PAGESIZE
#define LJ_PAGESIZE		4096
#endif

/* Check for minimum required compiler versions. */
#if defined(__GNUC__)
#if LJ_TARGET_X86
#if (__GNUC__ < 3) || ((__GNUC__ == 3) && __GNUC_MINOR__ < 4)
#error "Need at least GCC 3.4 or newer"
#endif
#elif LJ_TARGET_X64
#if __GNUC__ < 4
#error "Need at least GCC 4.0 or newer"
#endif
#elif LJ_TARGET_ARM
#if (__GNUC__ < 4) || ((__GNUC__ == 4) && __GNUC_MINOR__ < 2)
#error "Need at least GCC 4.2 or newer"
#endif
#elif LJ_TARGET_ARM64
#if __clang__
#if (__clang_major__ < 3) || ((__clang_major__ == 3) && __clang_minor__ < 5)
#error "Need at least Clang 3.5 or newer"
#endif
#else
#if (__GNUC__ < 4) || ((__GNUC__ == 4) && __GNUC_MINOR__ < 8)
#error "Need at least GCC 4.8 or newer"
#endif
#endif
#elif !LJ_TARGET_PS3
#if (__GNUC__ < 4) || ((__GNUC__ == 4) && __GNUC_MINOR__ < 3)
#error "Need at least GCC 4.3 or newer"
#endif
#endif
#endif

/* Check target-specific constraints. */
#ifndef _BUILDVM_H
#if LJ_TARGET_X64
#if __USING_SJLJ_EXCEPTIONS__
#error "Need a C compiler with native exception handling on x64"
#endif
#elif LJ_TARGET_ARM
#if defined(__ARMEB__)
#error "No support for big-endian ARM"
#endif
#if __ARM_ARCH_6M__ || __ARM_ARCH_7M__ || __ARM_ARCH_7EM__
#error "No support for Cortex-M CPUs"
#endif
#if !(__ARM_EABI__ || LJ_TARGET_IOS)
#error "Only ARM EABI or iOS 3.0+ ABI is supported"
#endif
#elif LJ_TARGET_ARM64
#if defined(__AARCH64EB__)
#error "No support for big-endian ARM64"
#endif
#if defined(_ILP32)
#error "No support for ILP32 model on ARM64"
#endif
#elif LJ_TARGET_PPC
#if defined(_SOFT_FLOAT) || defined(_SOFT_DOUBLE)
#error "No support for PowerPC CPUs without double-precision FPU"
#endif
#if !LJ_ARCH_PPC64 && LJ_ARCH_ENDIAN == LUAJIT_LE
#error "No support for little-endian PPC32"
#endif
#if LJ_ARCH_PPC64
#error "No support for PowerPC 64 bit mode (yet)"
#endif
#ifdef __NO_FPRS__
#error "No support for PPC/e500 anymore (use LuaJIT 2.0)"
#endif
#elif LJ_TARGET_MIPS
#if defined(_LP64)
#error "No support for MIPS64"
#endif
#endif
#endif

/* Enable or disable the dual-number mode for the VM. */
#if (LJ_ARCH_NUMMODE == LJ_NUMMODE_SINGLE && LUAJIT_NUMMODE == 2) || \
    (LJ_ARCH_NUMMODE == LJ_NUMMODE_DUAL && LUAJIT_NUMMODE == 1)
#error "No support for this number mode on this architecture"
#endif
#if LJ_ARCH_NUMMODE == LJ_NUMMODE_DUAL || \
    (LJ_ARCH_NUMMODE == LJ_NUMMODE_DUAL_SINGLE && LUAJIT_NUMMODE != 1) || \
    (LJ_ARCH_NUMMODE == LJ_NUMMODE_SINGLE_DUAL && LUAJIT_NUMMODE == 2)
#define LJ_DUALNUM		1
#else
#define LJ_DUALNUM		0
#endif

#if LJ_TARGET_IOS || LJ_TARGET_CONSOLE
/* Runtime code generation is restricted on iOS. Complain to Apple, not me. */
/* Ditto for the consoles. Complain to Sony or MS, not me. */
#ifndef LUAJIT_ENABLE_JIT
#define LJ_OS_NOJIT		1
#endif
#endif

/* 64 bit GC references. */
#if LJ_TARGET_GC64
#define LJ_GC64			1
#else
#define LJ_GC64			0
#endif

/* 2-slot frame info. */
#if LJ_GC64
#define LJ_FR2			1
#else
#define LJ_FR2			0
#endif

/* Disable or enable the JIT compiler. */
#if defined(LUAJIT_DISABLE_JIT) || defined(LJ_ARCH_NOJIT) || defined(LJ_OS_NOJIT) || LJ_FR2 || LJ_GC64
#define LJ_HASJIT		0
#else
#define LJ_HASJIT		1
#endif

/* Disable or enable the FFI extension. */
#if defined(LUAJIT_DISABLE_FFI) || defined(LJ_ARCH_NOFFI)
#define LJ_HASFFI		0
#else
#define LJ_HASFFI		1
#endif

#if defined(LUAJIT_DISABLE_PROFILE)
#define LJ_HASPROFILE		0
#elif LJ_TARGET_POSIX
#define LJ_HASPROFILE		1
#define LJ_PROFILE_SIGPROF	1
#elif LJ_TARGET_PS3
#define LJ_HASPROFILE		1
#define LJ_PROFILE_PTHREAD	1
#elif LJ_TARGET_WINDOWS || LJ_TARGET_XBOX360
#define LJ_HASPROFILE		1
#define LJ_PROFILE_WTHREAD	1
#else
#define LJ_HASPROFILE		0
#endif

#ifndef LJ_ARCH_HASFPU
#define LJ_ARCH_HASFPU		1
#endif
#ifndef LJ_ABI_SOFTFP
#define LJ_ABI_SOFTFP		0
#endif
#define LJ_SOFTFP		(!LJ_ARCH_HASFPU)

#if LJ_ARCH_ENDIAN == LUAJIT_BE
#define LJ_LE			0
#define LJ_BE			1
#define LJ_ENDIAN_SELECT(le, be)	be
#define LJ_ENDIAN_LOHI(lo, hi)		hi lo
#else
#define LJ_LE			1
#define LJ_BE			0
#define LJ_ENDIAN_SELECT(le, be)	le
#define LJ_ENDIAN_LOHI(lo, hi)		lo hi
#endif

#if LJ_ARCH_BITS == 32
#define LJ_32			1
#define LJ_64			0
#else
#define LJ_32			0
#define LJ_64			1
#endif

#ifndef LJ_TARGET_UNALIGNED
#define LJ_TARGET_UNALIGNED	0
#endif

/* Various workarounds for embedded operating systems or weak C runtimes. */
#if (defined(__ANDROID__) && !defined(LJ_TARGET_X86ORX64)) || defined(__symbian__) || LJ_TARGET_XBOX360 || LJ_TARGET_WINDOWS
#define LUAJIT_NO_LOG2
#endif
#if defined(__symbian__) || LJ_TARGET_WINDOWS
#define LUAJIT_NO_EXP2
#endif
#if LJ_TARGET_CONSOLE || (LJ_TARGET_IOS && __IPHONE_OS_VERSION_MIN_REQUIRED >= __IPHONE_8_0)
#define LJ_NO_SYSTEM		1
#endif

#if defined(LUAJIT_NO_UNWIND) || defined(__symbian__) || LJ_TARGET_IOS || LJ_TARGET_PS3 || LJ_TARGET_PS4
#define LJ_NO_UNWIND		1
#endif

/* Compatibility with Lua 5.1 vs. 5.2. */
#ifdef LUAJIT_ENABLE_LUA52COMPAT
#define LJ_52			1
#else
#define LJ_52			0
#endif

#endif
